### PR TITLE
Replace DBSelectionExtension with annotations

### LINKED ISSUE #1827: Replace DBSelectionExtension with annotations
The current DBSelectionExtension only provides class level management of what DB's to run or skip. Some test methods have been failing in a test, but options only include using **RunOnlyFor( dbs...)** or **SkipTestsFor** for all test cases.

Currently there are some tests that use the **@Disabled** annotation, but the tests are disabled for ALL DBTypes.

Create annotations that will allow running tests that can skip or run with individual test methods as well all tests in the test class.

### PR DESCRIPTION
**Pull request by blafond**
fixes #1827

### CODE REVIEW

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/GeneratedPropertySingleTableTest.java [hunk 1]
```diff
@@ -43,13 +42,14 @@
  * Test the @{@link Generated} annotation
  */
 @Timeout(value = 10, timeUnit = MINUTES)
-
+// It requires native queries, so it won't work for every db
+// Fails with SQLSERVER, ORACLE and DB2
+@RunDBType( value = POSTGRESQL )
```
Comment (DavideD) [inline hibernate-reactive-core/src/test/java/org/hibernate/reactive/GeneratedPropertySingleTableTest.java:47]: In all these cases, you can avoid specifying the attribute.
This will work:
````
@RunDBType( POSTGRESQL )
````

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/annotations/README.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
+These tests verify the new annotations intended to replace the **DBSelectionExtenstion** which only allows enabling/disabling some or all the tests within a test class.
+
+This extension feature didn't allow enabling/disabling at the test method level.
+
+The **RunDBType** and **SkipDBType** annotations are based off these ORM annotations and supporting utilities
+* [RequiresDialect](https://github.com/hibernate/hibernate-orm/blob/main/hibernate-testing/src/main/java/org/hibernate/testing/RequiresDialect.java)
+* [SkipForDialect](https://github.com/hibernate/hibernate-orm/blob/main/hibernate-testing/src/main/java/org/hibernate/testing/SkipForDialect.java)
```
Comment (DavideD) [inline hibernate-reactive-core/src/test/java/org/hibernate/reactive/annotations/README.md:7]: There's no need for this in the readme.
This should be mentioned in the JavaDoc

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/annotations/TestingUtil.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+import org.junit.platform.commons.support.AnnotationSupport;
+
+import static org.hamcrest.CoreMatchers.instanceOf;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+/**
+ * @author Steve Ebersole
+ * @see:  https://github.com/hibernate/hibernate-orm/hibernate-testing/src/main/java/org/hibernate/testing/orm/junit/TestingUtil.java
+ */
+public class TestingUtil {
```
Comment (DavideD) [inline hibernate-reactive-core/src/test/java/org/hibernate/reactive/annotations/TestingUtil.java:27]: Do we need this?

**Review by DavideD — CHANGES_REQUESTED** [review]
Thanks, I gave it a quick look and left a couple of minor changes requests. I will try to review it properly tomorrow.

I think I prefer the Jupiter convention for annotations.
I would prefer something like this:
```
@EnabledFor(POSTGRES)
@DisabledFor(DB2)
```

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/annotations/TestingUtil.java (hunk 3 above)
Comment (blafond) [reply hibernate-reactive-core/src/test/java/org/hibernate/reactive/annotations/TestingUtil.java:27]: Not the whole thing. I'll need one of the methods so I'll just move it to DBTypeFilterExtension.

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/annotations/README.md (hunk 2 above)
Comment (blafond) [reply hibernate-reactive-core/src/test/java/org/hibernate/reactive/annotations/README.md]: Done

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/GeneratedPropertySingleTableTest.java (hunk 1 above)
Comment (blafond) [reply hibernate-reactive-core/src/test/java/org/hibernate/reactive/GeneratedPropertySingleTableTest.java]: done

**Comment by blafond** [comment]
@DavideD I re-worked the logic to implement `ExecutionCondition` like you suggested.  It's lighter and cleaner and seems to work as expected.

The critical classes are: [DisableForDBTypeCondition](https://github.com/blafond/hibernate-reactive/blob/issue_1827/hibernate-reactive-core/src/test/java/org/hibernate/reactive/annotations/EnableForDBTypeCondition.java) and  [EnableForDBTypeCondition](https://github.com/blafond/hibernate-reactive/blob/issue_1827/hibernate-reactive-core/src/test/java/org/hibernate/reactive/annotations/EnableForDBTypeCondition.java)

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/annotations/tests/GuineaPig.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+import java.util.Objects;
+
+import jakarta.persistence.Entity;
+import jakarta.persistence.Id;
+import jakarta.persistence.Table;
+
+@Entity(name = "GuineaPig")
+@Table(name = "pig")
+public class GuineaPig {
```
Comment (DavideD) [inline hibernate-reactive-core/src/test/java/org/hibernate/reactive/annotations/tests/GuineaPig.java:16]: I guess this is a leftover from the previous PR

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/InternalStateAssertionsTest.java [hunk 5]
```diff
@@ -26,14 +25,16 @@
...
-import static org.hibernate.reactive.testing.DBSelectionExtension.runOnlyFor;
 import static org.hibernate.reactive.testing.ReactiveAssertions.assertThrown;
 
 /**
  * Checks that we throw the right exception when a session is shared between threads.
  * @see org.hibernate.reactive.common.InternalStateAssertions
  */
 @Timeout(value = 10, timeUnit = MINUTES)
+// These tests will fail before touching the database, so there is no reason
+// to run them on all databases
```
Comment (DavideD) [inline hibernate-reactive-core/src/test/java/org/hibernate/reactive/InternalStateAssertionsTest.java:36]: You could add this to the reason attribute, since we have one now

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/OrderQueriesTestBase.java [hunk 6]
```diff
@@ -189,7 +187,7 @@ public List<String> expectedQueries() {
 	}
 
 	// Should be able to group queries when they are ordered, and batch size is set
-	public static class OrderInsertsAndUpdatesWithBatchSizeTest extends OrderQueriesTestBase {
+	 public static class OrderInsertsAndUpdatesWithBatchSizeTest extends OrderQueriesTestBase {
```
Comment (DavideD) [inline hibernate-reactive-core/src/test/java/org/hibernate/reactive/OrderQueriesTestBase.java:190]: There's an extra white space

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/ReactiveMultitenantNoResolverTest.java [hunk 7]
```diff
@@ -51,13 +49,10 @@
  * @see ReactiveMultitenantTest
  */
 @Timeout(value = 10, timeUnit = MINUTES)
-
+// To check if we are using the right database we run native queries for PostgreSQL
+@EnableFor(POSTGRESQL)
```
Comment (DavideD) [inline hibernate-reactive-core/src/test/java/org/hibernate/reactive/ReactiveMultitenantNoResolverTest.java:53]: comment could go in the `reason` attribute of the annotation

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/ReactiveMultitenantTest.java [hunk 8]
```diff
@@ -38,15 +37,12 @@
  * on the database for the selected tenant.
  */
 @Timeout(value = 10, timeUnit = MINUTES)
-
+// To check if we are using the right database we run native queries for PostgreSQL
```
Comment (DavideD) [inline hibernate-reactive-core/src/test/java/org/hibernate/reactive/ReactiveMultitenantTest.java:40]: comment could go in the reason attribute of the annotation

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/configuration/ReactiveConnectionPoolTest.java [hunk 9]
```diff
@@ -39,23 +39,21 @@
...
 import static org.hibernate.reactive.containers.DatabaseConfiguration.getJdbcUrl;
-import static org.hibernate.reactive.testing.DBSelectionExtension.runOnlyFor;
 import static org.hibernate.reactive.testing.ReactiveAssertions.assertThrown;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 
 @ExtendWith(VertxExtension.class)
 @TestInstance(TestInstance.Lifecycle.PER_METHOD)
 @Timeout(value = 10, timeUnit = TimeUnit.MINUTES)
+// We use native queries, which may be different for other DBs
+@EnableFor(POSTGRESQL)
```
Comment (DavideD) [inline hibernate-reactive-core/src/test/java/org/hibernate/reactive/configuration/ReactiveConnectionPoolTest.java:49]: comment could go in the reason attribute of the annotation

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/configuration/ReactiveConnectionPoolTest.java [hunk 10]
```diff
@@ -39,23 +39,21 @@
...
 
-	@RegisterExtension
-	public DBSelectionExtension dbSelection = runOnlyFor( POSTGRESQL );
-
 	@RegisterExtension
 	public TestingRegistryExtension registryExtension = new TestingRegistryExtension();
 
 	@RegisterExtension
-	public RunTestOnContext testOnContext = new RunTestOnContext();
+	public static RunTestOnContext testOnContext = new RunTestOnContext();
```
Comment (DavideD) [inline hibernate-reactive-core/src/test/java/org/hibernate/reactive/configuration/ReactiveConnectionPoolTest.java:56]: Is this necessary? It doesn't seem related to this commit purpose

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/schema/SchemaUpdateCockroachDBTestBase.java [hunk 11]
```diff
@@ -44,7 +43,7 @@
 public abstract class SchemaUpdateCockroachDBTestBase extends BaseReactiveTest {
 
 	@Timeout(value = 10, timeUnit = MINUTES)
-
+	@EnableFor(COCKROACHDB)
```
Comment (DavideD) [inline hibernate-reactive-core/src/test/java/org/hibernate/reactive/schema/SchemaUpdateCockroachDBTestBase.java:46]: Just a curiosity, are this annotation inherited? Could I put them on the abstract class once instead?

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/types/BasicTypesAndCallbacksForAllDBsTest.java [hunk 12]
```diff
@@ -381,7 +377,7 @@ public void testBigIntegerType(VertxTestContext context) {
 	}
 
 	@Test
-	@Disabled // Fail for MSSQL because the value changes before it's saved on the db. This also fails for ORM
+	@DisableFor(value = MYSQL, reason = "fails because the value changes before it's saved on the db. This also fails for ORM")
```
Comment (DavideD) [inline hibernate-reactive-core/src/test/java/org/hibernate/reactive/types/BasicTypesAndCallbacksForAllDBsTest.java:380]: It seems this test now works for all databases (the comment says `MSSQL` but you've disabled it for `MYSQL`)

Comment (DavideD) [reply hibernate-reactive-core/src/test/java/org/hibernate/reactive/types/BasicTypesAndCallbacksForAllDBsTest.java:380]: I haven't tested it for all JDK though

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/annotations/tests/DisableForClassTest.java [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+import org.hibernate.reactive.annotations.DisableFor;
+import org.hibernate.reactive.containers.DatabaseConfiguration;
+
+import org.junit.jupiter.api.Test;
+
+import static org.hibernate.reactive.containers.DatabaseConfiguration.DBType.POSTGRESQL;
+import static org.junit.jupiter.api.Assertions.assertNotEquals;
+
+@DisableFor(value = POSTGRESQL, reason = "Reason #1")
+public class DisableForClassTest extends BaseReactiveTest {
```
Comment (DavideD) [inline hibernate-reactive-core/src/test/java/org/hibernate/reactive/annotations/tests/DisableForClassTest.java:19]: I don't think any of these tests need to extend from `BaseReactiveTest` (I might be wrong)

**Review by DavideD — CHANGES_REQUESTED** [review]
Thanks @blafond, it looks much nicer now.

There are some minor changes to apply, but looks good overall.

The test related to the annotation at a method level could be grouped in a single class:
```java
public class EnableForMethodTest  {

	@Test
	@EnableFor(value = POSTGRESQL)
	public void testEnable() {
		// Throw exception if this test is database is NOT POSTGRESQL
		assertEquals( POSTGRESQL, DatabaseConfiguration.dbType() );
	}

	@Test
	@EnableFor(value = POSTGRESQL)
	@EnableFor(value = MYSQL)
	public void testMultipleEnable() {
		// Throw exception if this test is run with POSTGRESQL or MYSQL database
		assertTrue( DatabaseConfiguration.dbType() == POSTGRESQL ||
							DatabaseConfiguration.dbType() == MYSQL );
	}

	@Test
	@EnableForGroup({
			@EnableFor(value = MYSQL),
			@EnableFor(value = POSTGRESQL)
	})
	public void testGroupEnable() {
		// Throw exception if this test is run with POSTGRESQL or MYSQL database
		assertTrue( DatabaseConfiguration.dbType() == POSTGRESQL ||
							DatabaseConfiguration.dbType() == MYSQL );
	}

       // Add the tests with disabled
}
```

This way we have less test classes around.
We probably could group togethere the others as well, but it would require to use subclasses and I'm not sure it's worth the effort.

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/schema/SchemaUpdateCockroachDBTestBase.java (hunk 11 above)
Comment (blafond) [reply hibernate-reactive-core/src/test/java/org/hibernate/reactive/schema/SchemaUpdateCockroachDBTestBase.java:46]: you are correct.  My previous annotations wouldn't allow it.

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/annotations/tests/DisableForMultipleClassTest.java [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+@DisableFor(value = MYSQL, reason = "Reason #1")
+@DisableFor(value = POSTGRESQL, reason = "Reason #2")
+public class DisableForMultipleClassTest extends BaseReactiveTest {
+
+	@Test
+	public void test() {
+		// Throw exception if this test is run with POSTGRESQL or MYSQL database
+		assertTrue( DatabaseConfiguration.dbType() != MYSQL && DatabaseConfiguration.dbType() != POSTGRESQL );
```
[the remark this answers is not in the document]
Comment (blafond) [reply hibernate-reactive-core/src/test/java/org/hibernate/reactive/annotations/tests/DisableForMultipleClassTest.java]: corrected

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/annotations/tests/DisableForMultipleMethodTest.java [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+
+public class DisableForMultipleMethodTest extends BaseReactiveTest {
+
+	@Test
+	@DisableFor(value = MYSQL, reason = "Reason #1")
+	@DisableFor(value = POSTGRESQL, reason = "Reason #2")
+	public void test() {
+		// Throw exception if this test is run with POSTGRESQL or MYSQL database
+		assertTrue( DatabaseConfiguration.dbType() != POSTGRESQL &&
+							DatabaseConfiguration.dbType() != MYSQL );
```
[the remark this answers is not in the document]
Comment (blafond) [reply hibernate-reactive-core/src/test/java/org/hibernate/reactive/annotations/tests/DisableForMultipleMethodTest.java:26]: corrected

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/annotations/tests/EnableForMultipleMethodTest.java [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+
+public class EnableForMultipleMethodTest {
+
+	@Test
+	@EnableFor(value = POSTGRESQL)
+	@EnableFor(value = MYSQL)
+	public void test() {
+		// Throw exception if this test is run with POSTGRESQL or MYSQL database
+		assertTrue( DatabaseConfiguration.dbType() == POSTGRESQL ||
+							DatabaseConfiguration.dbType() == MYSQL );
```
[the remark this answers is not in the document]
Comment (blafond) [reply hibernate-reactive-core/src/test/java/org/hibernate/reactive/annotations/tests/EnableForMultipleMethodTest.java]: corrected

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/annotations/tests/EnableForMultipleClassTest.java [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+
+@EnableFor(value = MYSQL)
+@EnableFor(value = POSTGRESQL)
+public class EnableForMultipleClassTest extends BaseReactiveTest {
+
+	@Test
+	public void test() {
+		// Throw exception if this test is run with POSTGRESQL or MYSQL database
+		assertTrue( DatabaseConfiguration.dbType() == POSTGRESQL ||
+							DatabaseConfiguration.dbType() == MYSQL );
```
[the remark this answers is not in the document]
Comment (blafond) [reply hibernate-reactive-core/src/test/java/org/hibernate/reactive/annotations/tests/EnableForMultipleClassTest.java]: corrected

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/configuration/ReactiveConnectionPoolTest.java (hunk 9 above)
Comment (blafond) [reply hibernate-reactive-core/src/test/java/org/hibernate/reactive/configuration/ReactiveConnectionPoolTest.java]: corrected

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/annotations/DisableFor.java [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+ * 		@DisableFor( value = POSTGRES, reason = "Reason #2")
+ * 		public void test(VertxTestContext context) {
+ *          ....
+ * 		}
+ * }
+ * }</pre>
+ *
+ *
+ * This annotation was based off of ORM's similar solution for Dialects:
+ *   @see org.hibernate.testing.orm.junit.SkipForDialect.java
```
[the remark this answers is not in the document]
Comment (blafond) [reply hibernate-reactive-core/src/test/java/org/hibernate/reactive/annotations/DisableFor.java]: removed

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/configuration/ReactiveConnectionPoolTest.java (hunk 10 above)
Comment (blafond) [reply hibernate-reactive-core/src/test/java/org/hibernate/reactive/configuration/ReactiveConnectionPoolTest.java]: no.. removed

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/types/BasicTypesAndCallbacksForAllDBsTest.java (hunk 12 above)
Comment (blafond) [reply hibernate-reactive-core/src/test/java/org/hibernate/reactive/types/BasicTypesAndCallbacksForAllDBsTest.java]: fixed

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/annotations/tests/DisableForClassTest.java (hunk 13 above)
Comment (blafond) [reply hibernate-reactive-core/src/test/java/org/hibernate/reactive/annotations/tests/DisableForClassTest.java]: correct. Removed the extension

**Comment by blafond** [comment]
Implemented your changes/suggestions and checked formatting as well in my changed files.  Found a few more unnecessary lines/comments.
